framework/test/MultipleListenersTest.cpp

```diff
@@ -0,0 +1,322 @@
+  }
+
+  void testListenerTypes()
+  {
+    class TestListener
+    {
+    public:
+      TestListener() : service_count(0), bundle_count(0), framework_count(0) {}
+
+      void serviceChanged(const ServiceEvent& evt)
+      {
+        US_TEST_OUTPUT( << "ServiceEvent: " << evt );
+        ++service_count;
+      }
+      void bundleChanged(const BundleEvent& evt)
+      {
+        US_TEST_OUTPUT( << "BundleEvent: " << evt );
+        ++bundle_count;
+      }
+      void frameworkChanged(const FrameworkEvent& evt)
+      {
+        US_TEST_OUTPUT( << "FrameworkEvent: " << evt );
+        ++framework_count;
+      }
+
+      std::vector<ListenerToken> tokens;
+      int service_count;
+      int bundle_count;
+      int framework_count;
+    };
+
+    auto f = FrameworkFactory().NewFramework();
+
+    f.Init();
+    auto fCtx = f.GetBundleContext();
+    TestListener tListen;
+
+    tListen.tokens.push_back(fCtx.AddServiceListener(&tListen, &TestListener::serviceChanged));
+    tListen.tokens.push_back(fCtx.AddServiceListener(&tListen, &TestListener::serviceChanged));
+    tListen.tokens.push_back(fCtx.AddBundleListener(&tListen, &TestListener::bundleChanged));
+    tListen.tokens.push_back(fCtx.AddBundleListener(&tListen, &TestListener::bundleChanged));
+    tListen.tokens.push_back(fCtx.AddFrameworkListener(&tListen, &TestListener::frameworkChanged));
+    tListen.tokens.push_back(fCtx.AddFrameworkListener(&tListen, &TestListener::frameworkChanged));
+
+    fCtx.RemoveListener(tListen.tokens[0]);
+    fCtx.RemoveListener(tListen.tokens[2]);
+    fCtx.RemoveListener(tListen.tokens[4]);
+    f.Start();
+
+    auto bundleA = testing::InstallLib(fCtx, "TestBundleA");
+    US_TEST_CONDITION_REQUIRED(bundleA, "Test for existing bundle TestBundleA")
+    bundleA.Start();
+
+    US_TEST_CONDITION(tListen.service_count == 1, "Test for number of times service listeners got triggered")
+    US_TEST_CONDITION(tListen.bundle_count == 4, "Test for number of times bundle listeners got triggered")
+    US_TEST_CONDITION(tListen.framework_count == 1, "Test for number of times framework listeners got triggered")
+
+    bundleA.Stop();
+    f.Stop();
+  }
+
+#ifdef US_ENABLE_THREADING_SUPPORT
+  void testConcurrent()
+  {
+    FrameworkFactory factory;
+    auto framework = factory.NewFramework();
+    framework.Init();
+    BundleContext fCtx = framework.GetBundleContext();
+
+    std::vector<std::future<ListenerToken>> futures;
+    std::vector<ListenerToken> tokens;
+    int count = 0;
+    const int num_listeners = 1001;
+    const int remove_count = num_listeners / 2;
+    auto add_listener = [&fCtx, &count]()
+    {
+      auto listener = [&count](const FrameworkEvent&){ ++count; };
+      auto token = fCtx.AddFrameworkListener(listener);
+      return token;
+    };
+
+    // 1. Add num_listeners number of listeners asynchronously.
+    // 2. Gather the resulting tokens.
+    // 3. Remove the first remove_count number of listeners using the tokens.
```

Comment:
This comment is not very helpful. Its describing what can be easily observed in the code. 
What would be most helpful to maintainers is to explain why this is being done.